src/Chapter3.hs

```diff
@@ -1006,6 +1128,99 @@ Implement data types and typeclasses, describing such a battle between two
 contestants, and write a function that decides the outcome of a fight!
 -}
 
+class Fighter a where
+  attack :: a -> Int
+  health  :: a -> Int
+  takeTurn :: a -> Fightable -> Fight
+  receiveDamage :: a -> Int -> a
+
+data Fightable =
+    FightHuman Human
+  | FightBeast Beast
+    deriving Show
+
+data Fight = Fight Fightable Fightable
+  deriving Show
+
+instance Fighter Fightable where
+  attack :: Fightable -> Int
+  attack (FightHuman human) = humanAtk human
+  attack (FightBeast beast) = beastAtk beast
+
+  health :: Fightable -> Int
+  health (FightHuman human) = humanHp human
+  health (FightBeast beast) = beastHp beast
+
+  receiveDamage :: Fightable -> Int -> Fightable
+  receiveDamage (FightHuman human) dmg = FightHuman (receiveDamage human dmg)
+  receiveDamage (FightBeast beast) dmg = FightBeast (receiveDamage beast dmg)
+
+  takeTurn :: Fightable -> Fightable -> Fight
+  takeTurn (FightHuman human) enemy = takeTurn human enemy
+  takeTurn (FightBeast beast) enemy = takeTurn beast enemy
+
+data Human = Human {
+  humanAtk :: Int,
+  humanHp :: Int,
+  humanDef :: Int,
+  humanActs :: [HumanAction]
+} deriving Show
+
+data HumanAction =
+    HumanAttack
+  | HumanDefend
+  | HumanHeal
+    deriving Show
+
+-- helper for cycling a fighters list of actions
+cycleOnce :: [a] -> [a]
+cycleOnce xs = take (length xs) . drop 1 . cycle $ xs
+
+instance Fighter Human where
+  attack = humanAtk
+  health = humanHp
+  receiveDamage self@(Human _ hp def _) dmg = self { humanHp = min (hp + def - dmg) hp}
+  takeTurn self enemy = case self of
+    Human atk hp def (act : _) -> case act of
+      HumanAttack -> Fight (receiveDamage enemy atk) (FightHuman self { humanActs = cycledActs self})
+      HumanDefend -> Fight enemy (FightHuman self { humanDef = def + 2, humanActs = cycledActs self })
+      HumanHeal -> Fight enemy (FightHuman self { humanHp = hp + 1, humanActs = cycledActs self})
+    _ -> Fight enemy (FightHuman self) -- dont need to cycle when acts is empty
+    where
+      cycledActs :: Human -> [HumanAction]
+      cycledActs = cycleOnce . humanActs
+
+data Beast = Beast {
+  beastAtk :: Int,
+  beastHp :: Int,
+  beastActs :: [BeastAction]
+} deriving Show
+
+instance Fighter Beast where
+  attack = beastAtk
+  health = beastHp
+  receiveDamage self@(Beast _ hp _) dmg = self { beastHp = hp - dmg}
+  takeTurn self enemy = case self of
+    Beast atk _ (act : _) -> case act of
+      BeastAttack -> Fight (receiveDamage enemy atk) (FightBeast self { beastActs = cycledActs self})
+      BeastFlee -> Fight enemy (FightBeast self { beastHp = 0, beastActs = cycledActs self })
+    _ -> Fight enemy (FightBeast self) -- dont need to cycle when acts is empty
+    where
+      cycledActs :: Beast -> [BeastAction]
+      cycledActs = cycleOnce . beastActs
+
+data BeastAction =
+    BeastAttack
+  | BeastFlee
+    deriving Show
+
+type Result = String
+
+resolveFight :: Fight -> Either Fight Fightable
```

Comment:
That is interesting, though the typeclass is not fully used in this solution. I think it would make sense to move `takeTurn` function out of the typeclass method. It could be a top-level function with constraints:

```haskell
takeTurn :: (FightTypeClass a, FightTypeClass b) => a -> b -> ...
```

and inside you could use the fact that both fighters have instances of this typeclass.

Also, `Fightable` data type seems redutand, you can try to use the mentioned logic to implement functions necessary for the fight 🙂